lib/spack/spack/cmd/checkbinaries.py

```diff
@@ -0,0 +1,247 @@
+import re
+import yaml
+
+import llnl.util.tty as tty
+
+import spack.cmd
+import spack.config
+from spack.error import SpackError
+from spack.spec import Spec
+from spack.util.executable import which
+
+
+description = "check if binaries need to be rebuilt"
+section = "developer"
+level = "long"
+
+
+_buildcache_path = os.path.join('build_cache')
+_pkg_match_regex = re.compile(r'packages[/\\]([^/\\]+)[/\\]package\.py')
+_pkg_yaml_match_regex = re.compile(r'<a href=[^>]+\>\s*([^<]+)\s*\<')
+
+
+def changed_package_files(args):
+    """Get list of changed package.py files in the Spack repository."""
+
+    git = which('git', required=True)
+
+    range = "{0}...".format(args.base)
+
+    git_args = [
+        # Add changed files committed since branching off of develop
+        ['diff', '--name-only', '--diff-filter=ACMR', range],
+        # Add changed files that have been staged but not yet committed
+        ['diff', '--name-only', '--diff-filter=ACMR', '--cached'],
+        # Add changed files that are unstaged
+        ['diff', '--name-only', '--diff-filter=ACMR'],
+    ]
+
+    changed = set()
+
+    for arg_list in git_args:
+        files = git(*arg_list, output=str).split('\n')
+
+        for f in files:
+            # Look only for modified package.py files
+            m = _pkg_match_regex.search(f)
+            if m:
+                changed.add(m.group(1))
+
+    return sorted(changed)
+
+
+def compute_local_hash(pkg_name, spec_yaml):
+    local_spec = Spec.from_yaml(spec_yaml)
+    tty.msg('local dag hash for %s: %s' % (pkg_name, local_spec.dag_hash()))
+    local_spec.concretize()
+    return local_spec.full_hash()
+
+
+def compute_remote_hash(pkg_name, spec_yaml):
+    parsed_yaml = yaml.load(spec_yaml)
+    if 'full_hash' in parsed_yaml:
+        return parsed_yaml['full_hash']
+
+    msg = 'Binary mirror yaml file is missing full_hash attribute, aborting'
+    raise SpackError(msg)
+
+
+def check_needs_rebuild(pkg_name, spec_yaml):
+    local_pkg_hash = compute_local_hash(pkg_name, spec_yaml)
+    remote_pkg_hash = compute_remote_hash(pkg_name, spec_yaml)
+
+    tty.msg('  %s' % pkg_name)
+    tty.msg('    local hash: %s' % local_pkg_hash)
+    tty.msg('    remote hash: %s' % remote_pkg_hash)
+
+    if local_pkg_hash != remote_pkg_hash:
+        return True
+
+    return False
+
+
+def parse_yaml_list(index_html_contents):
+    result = []
+    match_list = _pkg_yaml_match_regex.findall(index_html_contents)
+    if match_list:
+        result = match_list
+    return result
+
+
+def get_mirror_rebuilds(mirror_name, mirror_url, pkg_change_set):
+    tty.msg('Checking for built specs on %s' % mirror_name)
+
+    # First fetch the index.html
+    index_path = os.path.join(mirror_url, _buildcache_path, 'index.html')
+    index_contents = None
+
+    try:
+        url = urlopen(index_path)
+    except URLError as urlErr:
+        tty.error('Unable to open url %s' % index_path)
+        tty.error(urlErr)
+        return None
```

Comment:
These `return None`s should probably raise some type of `SpackError`.  Any SpackErrorr raised will end the command and print the message from the error, or a stack trace in debug mode.

Reply:
Sounds good.